### PR TITLE
implement ring attetion with zigzag layout

### PR DESCRIPTION
**Pull request by haok1402**
Changes:

1. use zigzag layout to ensure causal load balance under context parallelism
2. overlap flash attention compute over the current block with the send-recv of other KV blocks
3. apply torch compile to the combination of lse and out

---

Qwen3-30b-a3b, cp4-s32k, one layer:

- Total time: 13.87 ms → 8.30 ms (1.67× faster)
- Forward: 4.14 ms → 2.47 ms (1.68×)
- Backward: 9.72 ms → 5.83 ms (1.67×)

---

On the main branch, the workload with naive ring attention is imbalanced. At earlier ranks, NCCL stream took 83.1% of the time, waiting for the last rank to complete, which spends 95.2% on the default stream (i.e. compute), because it can attend to all previous positions.

image

On the current branch, the split between NCCL stream and the default stream is more even, with the transfer of KV blocks happening while the FA4 kernels run.

image

### CODE REVIEW

File: pithtrain/models/deepseek_v2_lite.py [hunk 1]
```diff
@@ -714,11 +715,23 @@ def forward(
...
+        block = seq_len // 2
+        global_seq_len = seq_len * self.cp_size
+        front_start = self.cp_rank * block
+        back_start = (2 * self.cp_size - self.cp_rank - 1) * block
+        position_ids = torch.cat(
+            [
+                torch.arange(front_start, front_start + block, device=hidden_states.device),
+                torch.arange(back_start, back_start + block, device=hidden_states.device),
+            ]
+        )
```
[the remark this answers is not in the document]
Comment (haok1402) [reply pithtrain/models/deepseek_v2_lite.py:730]: not really, after concat, these tensors 64 KB int64 tensor... i don't see it being a heavy operation ; plus, it is created once per model forward.

File: pithtrain/models/qwen3_30b_a3b.py [hunk 2]
```diff
@@ -765,12 +766,21 @@ def forward(
...
+        block = seq_len // 2
+        global_seq_len = seq_len * self.cp_size
+        front_start = self.cp_rank * block
+        back_start = (2 * self.cp_size - self.cp_rank - 1) * block
+        position_ids = torch.cat(
+            [
+                torch.arange(front_start, front_start + block, device=hidden_states.device),
+                torch.arange(back_start, back_start + block, device=hidden_states.device),
+            ]
         )
```
[the remark this answers is not in the document]
Comment (haok1402) [reply pithtrain/models/qwen3_30b_a3b.py:781]: not really, after concat, these tensors 64 KB int64 tensor... i don't see it being a heavy operation ; plus, it is created once per model forward.

File: pithtrain/operators/ring_attention.py [hunk 3]
```diff
@@ -1,191 +1,306 @@
...
+    grad_work: Optional[List[Work]] = None
+
+    for step in range(cp_size):
+        if step + 1 < cp_size:
+            next_k, next_v, kv_work = post(k, v)
+        if step == 0:
+            dq_step, dk_step, dv_step = bwd(q, k, v, out, dout, lse, causal=True)
+            dq = dq_step.to(torch.float32)
+            dk = dk_step.to(torch.float32)
+            dv = dv_step.to(torch.float32)
```
[the remark this answers is not in the document]
Comment (haok1402) [reply pithtrain/operators/ring_attention.py:225]: wrong. gradients should be accumulated in fp32.

File: pithtrain/operators/ring_attention.py [hunk 4]
```diff
@@ -1,191 +1,306 @@
...
 
-def _online_softmax_combine(out1, lse1, out2, lse2):
-    """Combine two partial attention outputs via online softmax rescaling.
+def wait_ring(work: List[Work]) -> None:
+    for req in work:
+        req.wait()
+
 
-    out: [B, S, H, D],  lse: [B, H, S]  (float32).
+@torch.compile(fullgraph=True)
```
Comment (MasterJH5574) [inline pithtrain/operators/ring_attention.py:102]: Do we need the decorator here? Isn't the ring attention always run under an outside `@torch.compile` decorator?

Comment (MasterJH5574) [reply pithtrain/operators/ring_attention.py:102]: Maybe need to quickly check the kernel used here without the decorator.

File: pithtrain/operators/ring_attention.py [hunk 5]
```diff
@@ -1,191 +1,306 @@
...
+    dst = get_global_rank(cp_group, (cp_rank + 1) % cp_size)
+    src = get_global_rank(cp_group, (cp_rank - 1) % cp_size)
+    block = q.shape[1] // 2
+    q_back = q[:, block:]
+
+    def fwd(q, k, v, causal=False):
+        return _flash_attn_fwd(q, k, v, softmax_scale=sm_scale, causal=causal, return_lse=True)
+
+    def post(k, v, k_recv=None, v_recv=None):
+        return post_ring_kv(k, v, cp_group, dst, src, k_recv, v_recv)
```
Comment (MasterJH5574) [inline pithtrain/operators/ring_attention.py]: These two functions don't seem necessary. We can inline them.

File: pithtrain/operators/ring_attention.py [hunk 6]
```diff
@@ -1,191 +1,306 @@
...
+    src = get_global_rank(cp_group, (cp_rank - 1) % cp_size)
+    block = q.shape[1] // 2
 
+    def bwd(q, k, v, out, dout, lse, causal=False):
+        return _flash_attn_bwd(q, k, v, out, dout, lse, softmax_scale=sm_scale, causal=causal)
 
-class RingAttentionFunc(torch.autograd.Function):
-    """Causal ring attention with flash-attn kernels."""
+    def post(k, v, k_recv=None, v_recv=None):
+        return post_ring_kv(k, v, cp_group, dst, src, k_recv, v_recv)
```
Comment (MasterJH5574) [inline pithtrain/operators/ring_attention.py]: Ditto

File: pithtrain/operators/ring_attention.py (hunk 4 above)
Comment (haok1402) [reply pithtrain/operators/ring_attention.py:102]: > Isn't the ring attention always run under an outside `@torch.compile`

No. We currently unwrap the `torch.compile` if ring attention is used.

https://github.com/mlc-ai/Pith-Train/blob/c76a5916b1c8e4aa1732f8a617184904119a2deb/pithtrain/models/qwen3_30b_a3b.py#L486-L489

We apply the `torch.compile(fullgraph=True)` decorator, mainly to fuse the kernels in `combine_partial`. 

image

image

We do see slight reduction with the latency,

- before this compile : fwd 3.004 ms, bwd 6.174 ms, step 9.178 ms
- after : fwd 2.467 ms, bwd 5.833 ms, step 8.300 ms

Comment (haok1402) [reply pithtrain/operators/ring_attention.py:102]: > > Isn't the ring attention always run under an outside `@torch.compile`
> 
> No. We currently unwrap the `torch.compile` if ring attention is used.
> 
> https://github.com/mlc-ai/Pith-Train/blob/c76a5916b1c8e4aa1732f8a617184904119a2deb/pithtrain/models/qwen3_30b_a3b.py#L486-L489

@MasterJH5574 I think `torch.compile` can overall be applicable with ring attention. We should have a follow up PR to address this.

File: pithtrain/operators/ring_attention.py (hunk 5 above)
Comment (haok1402) [reply pithtrain/operators/ring_attention.py]: fixed.

File: pithtrain/operators/ring_attention.py (hunk 6 above)
Comment (haok1402) [reply pithtrain/operators/ring_attention.py]: fixed.

**Review by MasterJH5574 — APPROVED** [review]
Thanks @haok1402!
